### ISSUE TITLE
Memory leak

### ISSUE THREAD

**Issue by infinite3219** [astonbitecode/j4rs#136]
Hello,

I'm noticing memory leaks when using j4rs to call a Java function from Rust. The Rust function passes a string, the Java function takes that string and does some operations on it and returns a String as a result.

The program basically works as:
setup.rs to initialize the JVM once for entire program

```
    JvmBuilder::new()
        .classpath_entries(entries)
        .build()
```

Then the Java function is invoked as follows (I've omitted a lot of stuff around it for simplicity)

```
        tokio::select! {
              tokio::task::spawn_blocking(move || {
                    result = call_java_function_wrapper(rust_string)
               })

```

```
  pub fn call_java_function_wrapper(hex_message: &str) -> J4RsResult<String> {
        // make current thread attach to JVM
        let jvm: Jvm = Jvm::attach_thread().unwrap();

        // Convert the hex message to an argument
        let arg = InvocationArg::try_from(hex_message)?;

        // Call the static decode method
        let result = jvm.invoke_static("com.abc.sim.Decoder", "decodeMessage", &[&arg])?;
        // Convert the result back to a Rust String
        jvm.to_rust(result)
    }

```

I tried passing the string to Java via reference and move, but that made no difference.

I did some further experiments where I didn't actually call the Java function, but instead just did the ```        let jvm: Jvm = Jvm::attach_thread().unwrap();``` and noticed that this alone was sufficient to cause memory growth (when done in a loop). The same function that returns a string without the ```attach_thread``` doesn't cause this memory growth in a loop.

Perhaps I'm missing something to deallocate something at the end of this function. Please let me know if you have any thoughts.

**Comment by astonbitecode** [comment]
Hi,

which version do you use? There was a leak that was fixed some time ago in the context of [this issue](https://github.com/astonbitecode/j4rs/issues/40). For that, the fix version was v0.15.0.

**Comment by infinite3219** [comment]
I was using: 

j4rs = "0.21.0"

**Comment by astonbitecode** [comment]
How did you notice the memory leak?
I have tested your case with [ASAN]( https://doc.rust-lang.org/unstable-book/compiler-flags/sanitizer.html) running the use case you describe:

```rust
for i in 0.. 10000000 {
        tokio::task::spawn_blocking(move || {
            if i % 100000 == 0 {
                println!("JVM {i}");
            }
            let _ = Jvm::attach_thread().unwrap();
            // Uncomment this and the sanitizer fails
            // let b = Box::new(i as f64);
            // Box::leak(b);
            if i % 100000 == 0 {
                println!("JVM {i} ok");
            }
        });

// Uncomment this and there is no memory increase
//let millis = std::time::Duration::from_nanos(100);
//thread::sleep(millis);
   
}
```
And no leaks were reported. For testing, you may uncomment the leak in the example above and the sanitizer will report the error.

Indeed the memory vastly increases when the application starts, but after some time it gets stabilized. Maybe this has to do with the fact that the parallelization we get with tokio is performing really well and the resources needed to do the attach are increasing. 
For example, if you add a small thread sleep for each loop, the memory is not increased at all.
```rust
let millis = std::time::Duration::from_nanos(100);
thread::sleep(millis);
```

**Comment by infinite3219** [comment]
I re-did the test and it turns out the leak doesn't happen unless you actually call the ```jvm.invoke_static```. The way I noticed the leak was to run top and look at RSS memory on Linux or MEM column on macOS. You can see a rapid increase of memory that keeps going up (I have run it until it hit ~20GB of memory).

The same Java function doesn't cause any issues if I use the ```jni-rs``` crate. The Java side does allocate memory for string manipulations. How does this memory get released? Is there garbage collection that happens as part of the JVM or is it supposed to release the memory as soon the JNI function call ends? I suspect it is the latter because there is no memory growth at all with ```jni-rs```.

**Comment by astonbitecode** [comment]
You are right, there was a problem while calling the `invoke_static`, thanks for the finding!

I was able to debug it and found a local ref that was not being deallocated. 

My tests are passing with this change. I would be grateful if you also give it a try using the master.

Thanks once again.

**Comment by astonbitecode** [comment]
@infinite3219, dou you maybe have some update on this?

**Comment by astonbitecode** [comment]
I guess this can be closed.
